modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/ByteBufBytesReference.java

```diff
@@ -1,104 +0,0 @@
-/*
- * Licensed to Elasticsearch under one or more contributor
- * license agreements. See the NOTICE file distributed with
- * this work for additional information regarding copyright
- * ownership. Elasticsearch licenses this file to you under
- * the Apache License, Version 2.0 (the "License"); you may
- * not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *    http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing,
- * software distributed under the License is distributed on an
- * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
- * KIND, either express or implied.  See the License for the
- * specific language governing permissions and limitations
- * under the License.
- */
-package org.elasticsearch.transport.netty4;
-
-import io.netty.buffer.ByteBuf;
-import org.apache.lucene.util.BytesRef;
-import org.elasticsearch.common.bytes.AbstractBytesReference;
-import org.elasticsearch.common.bytes.BytesReference;
-import org.elasticsearch.common.io.stream.StreamInput;
-
-import java.io.IOException;
-import java.io.OutputStream;
-import java.nio.charset.StandardCharsets;
-
-final class ByteBufBytesReference extends AbstractBytesReference {
-
-    private final ByteBuf buffer;
-    private final int length;
-    private final int offset;
-
-    ByteBufBytesReference(ByteBuf buffer, int length) {
-        this.buffer = buffer;
-        this.length = length;
-        this.offset = buffer.readerIndex();
-        assert length <= buffer.readableBytes() : "length[" + length +"] > " + buffer.readableBytes();
-    }
-
-    @Override
-    public byte get(int index) {
-        return buffer.getByte(offset + index);
-    }
-
-    @Override
-    public int getInt(int index) {
-        return buffer.getInt(offset + index);
-    }
-
-    @Override
-    public int indexOf(byte marker, int from) {
-        final int start = offset + from;
-        return buffer.forEachByte(start, length - start, value -> value != marker);
```

Comment:
I'm not sure removing this optimisation in particular is a good idea. This saved quite a bit of CPU for finding the EOLs in bulk requests. I think the composite byte buffer will behave a lot worse here? (especially now that bulk requests aren't copied to a single `byte[]` any more and we parse them straight from the `ByteBuf` we get from the HTTP layer.

Reply:
This commit implements that optimization for both the `CompositeBytesReference` and `ByteBufferReference`.